### PR TITLE
STM32 Flash fix and optimizations

### PR DESCRIPTION
**Pull request by erwango**
Fix in STM32l4 driver that could be vulnerable to unaligned memory access if data is not aligned in flash_write operation.

Additionally, rework flash registers definitions for STM32F1/STM32F3 (missing volatile instruction).

### CODE REVIEW

File: drivers/flash/flash_stm32l4x.c [hunk 1]
```diff
@@ -149,9 +149,15 @@ int flash_stm32_write_range(struct device *dev, unsigned int offset,
...
 {
 	int i, rc = 0;
+	u32_t *tmp = (u32_t *)data;
+	u64_t dword;
 
 	for (i = 0; i < len; i += 8, offset += 8) {
-		rc = write_dword(dev, offset, ((const u64_t *) data)[i>>3]);
+		/* Use local variable to access 'data' word by word */
+		/* to avoid errors in case 'data' is unaligned */
+		dword = ((u64_t)(tmp + (i >> 2))[1] << 32) +
```
Comment (galak) [inline drivers/flash/flash_stm32l4x.c]: Would UNALIGNED_GET be useful here?

File: drivers/flash/flash_stm32.c [hunk 2]
```diff
@@ -210,11 +210,9 @@ static int flash_stm32_write_protection(struct device *dev, bool enable)
...
 			regs->keyr = FLASH_KEY1;
-#ifdef CONFIG_SOC_SERIES_STM32F3X
-			/* On STM32F3 series some time is requested */
-			/* before writing again on key register */
-			flash_stm32_wait_flash_idle(dev);
-#endif /* CONFIG_SOC_SERIES_STM32F3X */
+			/* Order of operations is critical here */
+			/* Add a barrier to avoid compiler optimisations */
+			__ISB();
 			regs->keyr = FLASH_KEY2;
```
Comment (aurel32) [inline drivers/flash/flash_stm32.c]: I am not sure that you need an instruction barrier, the flash registers are in a strongly ordered region, so the CPU should already ensure that the two writes actually reach the registers.

OTOH you might want a compiler barrier, but that should not be needed if the registers are declared `volatile`. And I just realized that all the flash registers are declared as volatile, except on the STM32F1x and STM32F3x. Could it be the real issue?

File: drivers/flash/flash_stm32l4x.c [hunk 3]
```diff
@@ -149,9 +149,15 @@ int flash_stm32_write_range(struct device *dev, unsigned int offset,
...
+			 (u64_t)(tmp + (i >> 2))[0];
+		rc = write_dword(dev, offset, dword);
 		if (rc < 0) {
```
Comment (aurel32) [inline drivers/flash/flash_stm32l4x.c]: That part looks all good.

File: drivers/flash/flash_stm32.c (hunk 2 above)
Comment (erwango) [reply drivers/flash/flash_stm32.c]: >  Could it be the real issue?

This sounds more realistic indeed. I'll have a try.

File: drivers/flash/flash_stm32l4x.c (hunk 1 above)
Comment (erwango) [reply drivers/flash/flash_stm32l4x.c]: Let me have a test

File: drivers/flash/flash_stm32.c (hunk 2 above)
Comment (erwango) [reply drivers/flash/flash_stm32.c]: That was it, thanks for the hint!

File: drivers/flash/flash_stm32l4x.c (hunk 1 above)
Comment (erwango) [reply drivers/flash/flash_stm32l4x.c]: Seems like UNALIGNED_GET doesn't like the expression it should be applied to.

Comment (aurel32) [reply drivers/flash/flash_stm32l4x.c]: `UNALIGNED_GET` needs a pointer in argument, so you need to use UNALIGNED_GET((const u64_t *) data + (i >> 3))``

Comment (erwango) [reply drivers/flash/flash_stm32l4x.c]: Right! fixed,  thanks again.

**Review by aurel32 — APPROVED** [review]
Thanks for the update, LGTM.

**Pull request merged**